### ISSUE TITLE
Rendering problem when removing and adding entries

### ISSUE THREAD

**Issue by waver55** [stefanuebe/vaadin_fullcalendar#135]
Hi,
I just tried to update to the 4.0.3 version, but I am experiencing a strange behavior.
When an Entry is dropped or resized, I write the changes back to the database and reload all events in the current time range, something like this:
`calendar.removeAllEntries();`
`Entry[] entries = loadEventsFromBackend();`
`calendar.addEntries(entries);`

This removes all entries from the calendar, the new added entries do not show up, and I'm left with an empty calendar. If I refresh the page, all entries show up again. 
I haven't found the time to analyze more, but I can provide more information if needed.

**Comment by stefanuebe** [comment]
Might be, that the internal "change" caches clash here. I'll try to check that use case nearby. With the upcoming 4.1 it might be not necessay anymore to to it that way in this case, though

**Comment by aetasoul** [comment]
Hi, 
in my code I do the same thing as you, let me check it. I'll upgrade it later and test it with the 4.0.3.

**Comment by aetasoul** [comment]
Hi @waver55 
this happend in with which view type? I use the resource_timeline_year and all works fine.

My code remove all the entries from the calendar, refetch from the database the entries and readd to the calendar the entries and the resources.

All show up correctly. Could be a problem with you time range?

Can you check the entries array if it is empty? ( just to be sure the refetch code from the db is ok)
Another thing to check is to convert you array to iterable before adding the entries back to the calendar, in my code I manage the entries as List:

```java
calendar.removeAllEntries();
List<Entry> entries = loadEventsFromBackend();
calendar.addEntries((Iterable<Entry>) entries);
```

Can you try this please? Just to isolate the problem.

**Comment by aniwareshared** [comment]
Hi!! I have the same issues as @waver55  wrote about.
If use removeAllEntries() and then addEntries the display becomes unstable.. sometimes it works and everything is ok, but for the most part the calendar is empty (although elements are added in the calendar object)
p.s. added entries like ArrayList

**Comment by aetasoul** [comment]
Hi @aniwareshared 
the questions above are also valid for you

**Comment by aniwareshared** [comment]
easy way to get this issues.. 
add to addDatesRenderedListener -> remove all entries and add again 
then -> click button "Today"

**Comment by aniwareshared** [comment]
by default use TIME_GRID_WEEK. Entries array not empty.

**Comment by aetasoul** [comment]
I have to test it.

In my project I have something like this:

```java

public class MyClass{

  public MyClass(){
    // Create calendar etc

    calendar.addEntryDroppedSchedulerListener((EntryDroppedSchedulerEvent e) -> {
      // Do stuff
      UpdateView();
    });

  calendar.addEntryResizedListener((EntryResizedEvent e) -> {
      // Do stuff
      UpdateView();
    });

  // Add all the other events that I need to handle, and at the end call the UpdateView
  }

  public void UpdateView(){
    calendar.removeAllEntries();
    List<Entry> entries = loadEventsFromBackend();
    calendar.addEntries((Iterable<Entry>) entries);
  }
}
```

**Comment by waver55** [comment]
Hey, I'm also using TIME_GRID_WEEK, and I am not using the Scheduler Version. If I find time I will try to debug the issue, maybe I will get to it later today.

**Comment by aniwareshared** [comment]
FYI: I also use calendar without scheduler

**Comment by aetasoul** [comment]
Perfect thank, I'll also test it on the demo and try to replicate

**Comment by waver55** [comment]
I debugged the code and noticed a few things.
First removeAllEntries calls the triggerClientSideUpdate() method twice, tho i do not think this causes any issues.
The changes from removeAllEntries get applied when addEntries is called, so it seems the code`  this.getElement().getNode().runWhenAttached((ui) ->`  in the triggerClientSideUpdate method only runs a single time, even tho two methods get called. Now the method merges the remove call and the add call, and `  this.tmpAdd.removeAll(this.tmpRemove);` now removes all the added Entries again, which is why the calendar is empty.
Maybe its possible do enforce the delete? Or to write some logic that remembers the order in which the commands came in?

**Comment by stefanuebe** [comment]
Just thinking loud: It might make sense to move the tmpXYZ changes to the respective methods instead of doing it at the request end

* when calling add, it should it should remove the item from tmpRemove. 
* when calling remove, it should add the item to the tmpRemove list and remove it from tmpAdd and tmpUpdate.

Additional necessary changes
* update needs to check, if the item is known to the client and is in the entries map and is not registered as "to add": then it might be added to tmpUpdate

**Comment by waver55** [comment]
When the item will be removed from tmpRemove on the add call it should be checked if the entries are equal in all fields, otherwise a changed startTime or color from the new entry with the same id would not be applied.

**Comment by stefanuebe** [comment]
If it is in the tmpAdd set, it will be sent fully to the client. So it should recognize all changes.

**Comment by waver55** [comment]
Depending on the exact implementation it could be problematic:
removeAllEntries -> tmpRemove will be filled
addAllEntrie with all the same Ids but different value -> will remove everything from tmpRemove (and maybe add to the tmpAdd)
Either the values will not change at all, since both lists are empty, or the entries will be added a second time(If its even possible to add the entries again with the same id), or does the calendar automatically update the values if the same id is added again? 
Otherwise there would be a need to update the entity if its in tmpRemove and then added again with different values maybe?

I hope my description is clear enoug 😄

**Comment by stefanuebe** [comment]
After a 2nd reading I get your point. In short, the readded items might result in duplicates by id on the client. 

Then, if the tmpRemove and tmpAdd do not interfere with each other, but simply do it in a way, that we 
1. send removed
2. send added
3. send updated

update is a bit a special case, but theoretically a different point.

**Comment by stefanuebe** [comment]
Another good point. For that I could add a check, if the id is still registered on "trigger" time as it is already done with the update part.

**Comment by stefanuebe** [comment]
```
protected void executeClientSideUpdate() {
    Map<String, T> entriesMap = getEntriesMap();

    if (resendAll) {
        tmpRemove.addAll(entriesMap.values());
        tmpAdd.addAll(entriesMap.values());
    }

    // items that are not yet on the client need no update
    tmpUpdate.removeAll(tmpAdd);

    // items to be deleted, need not to be added or updated on the client
    tmpUpdate.removeAll(tmpRemove);

    convertItemsAndSendToClient("removeEvents", tmpRemove, item -> {
        // only send some json to delete, if the item has already been sent to the client once
        if (item.isKnownToTheClient()) {
            JsonObject jsonObject = item.toJsonWithIdOnly();
            item.setKnownToTheClient(false);
            return jsonObject;
        }
        return null;
    });

    convertItemsAndSendToClient("addEvents", tmpAdd, item -> {
        if (entriesMap.containsKey(item.getId())) { // prevent accidentally adding items, that have been removed after adding
            JsonObject json = item.toJson(false);
            item.setKnownToTheClient(true);
            item.clearDirtyState();
            return json;
        }
        return null;
    });

    convertItemsAndSendToClient("updateEvents", tmpUpdate, item -> {
        String id = item.getId();
        if (item.isDirty() && entriesMap.containsKey(id)) {
            JsonObject json = item.toJson(true);
            item.clearDirtyState();
            return json;
        }
        return null;
    });


    tmpAdd.clear();
    tmpUpdate.clear();
    tmpRemove.clear();

    updateRegistered = false;

    // if there is a request to resend all, it must be removed here
    resendAll = false;
}

```

**Comment by stefanuebe** [comment]
Btw since the 4.1.0 with the new entry providers is about to be published, I will publish that fix within that version.

**Comment by stefanuebe** [comment]
The branch `v4.1.0_dataprovider` contains the fixed, some unit tests regarding that issue and also some test views. Can someone please verify, that the issue will be fixed with all that stuff?

Tests will be part of the addon module. Test views can be seen in the demo app in the "hidden" test section at `http://localhost:8080/test`

**Comment by aetasoul** [comment]
I'll try to check it later

**Comment by aetasoul** [comment]
Seems to be all ok, I'd like to wait for the @waver55 or @aniwareshared comments, for me you can release it

**Comment by stefanuebe** [comment]
Released version 4.1.0. The new EagerInMemoryEntryProvider, who inherits the previous FullCalendar entries "cache", also contains the fixed version of pushing the data.

Nevertheless, I recommend to switch to one of the other EntryProviders (as they seem to be more fitting for your use case).

**Comment by stefanuebe** [comment]
PS: due to the huge amount of comments, if there still is a problem, please open a new issue referring to this one.
